ee/performance/stats.go

```diff
@@ -56,51 +58,113 @@ func ProcessStatsForPid(ctx context.Context, pid int) (*PerformanceStats, error)
 		return nil, fmt.Errorf("getting process handle for pid %d: %w", pid, err)
 	}
 
+	var memStats runtime.MemStats
+	runtime.ReadMemStats(&memStats)
+
+	ps, memInfo, err := statsForProcess(ctx, proc)
+	if err != nil {
+		return nil, fmt.Errorf("gathering stats for process: %w", err)
+	}
+
+	ps.MemInfo.NonGoMemUsage = nonGoMemUsage(memInfo, &memStats)
+	ps.MemInfo.GoMemUsage = goMemUsage(&memStats)
+	ps.MemInfo.HeapTotal = heapTotal(&memStats)
+
+	// Record stats
+	observability.GoMemoryUsageGauge.Record(ctx, int64(ps.MemInfo.GoMemUsage))
+	observability.NonGoMemoryUsageGauge.Record(ctx, int64(ps.MemInfo.NonGoMemUsage))
+	observability.MemoryPercentGauge.Record(ctx, int64(ps.MemInfo.MemPercent))
+	observability.CpuPercentGauge.Record(ctx, int64(ps.CPUPercent))
+	observability.RSSHistogram.Record(ctx, int64(ps.MemInfo.RSS))
+
+	return ps, nil
+}
+
+func CurrentProcessChildStats(ctx context.Context) ([]*PerformanceStats, error) {
+	pid := os.Getpid()
+	return ChildProcessStatsForPid(ctx, int32(pid))
+}
+
+func ChildProcessStatsForPid(ctx context.Context, pid int32) ([]*PerformanceStats, error) {
+	proc, err := process.NewProcessWithContext(ctx, pid)
+	if err != nil {
+		return nil, fmt.Errorf("getting process handle for pid %d: %w", pid, err)
+	}
+
+	childProcesses, err := proc.ChildrenWithContext(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("getting child processes for pid %d: %w", pid, err)
+	}
+
+	stats := make([]*PerformanceStats, 0)
+	for _, childProcess := range childProcesses {
+		ps, _, err := statsForProcess(ctx, childProcess)
+		if err != nil {
+			continue
+		}
+		stats = append(stats, ps)
+
+		if strings.Contains(ps.Cmdline, "osquery") {
+			observability.OsqueryRssHistogram.Record(ctx, int64(ps.MemInfo.RSS))
```

Comment:
Do we want to keep these as a side effect of stats collection? It does seem clean here but would feel a little more flexible to me if the caller was in charge of setting this stuff after the stats collection took place

Reply:
My thought was that any time we're doing the work to gather this data, we'd want to emit the metrics associated with it too. I am pretty sure I read somewhere in the [otel spec](https://opentelemetry.io/docs/specs/otel/metrics/api/) that instruments should only ever be used in one place -- I can't find it there now, but that's also why I originally wanted to record the metrics as close to where we gather them as possible. (Plus we'd need to rework the log checkpointer to conditionally record metrics.)

I agree it feels a little weird/not quite correct as a side effect, though...at the very least, I can document these side effects better.